### PR TITLE
YJIT: Add known_* helpers for Type

### PR DESCRIPTION
**Pull request by jhawthorn**
This adds a few helpers to Type which all return `Options` representing what is known, from a Ruby perspective, about the type.

This includes:
* known_class_of: If known, the class represented by this type
* known_value_type: If known, the T_ value type
* known_exact_value: If known, the exact VALUE represented by this type (currently this is only available for true/false/nil)
* known_truthy: If known, whether or not this value evaluates as true (not false or nil)

The goal of this is to abstract away the specifics of the mappings between types wherever possible from the codegen. For example previously by introducing `Type::CString` as a more specific version of `Type::TString`, uses of `Type::TString` in codegen needed to be updated to check either case. Now by using known_value_type, at least in theory we can introduce new types with minimal (if any) codegen changes.

I think rust's Option type allows us to represent this uncertainty fairly well, and should help avoid mistakes, and the matching using this turned out pretty cleanly.

This PR includes a commit using each of these helpers as a demonstration (other than `known_exact_value`, which I didn't find a place to use 😅):
* I really like how it removed a bunch of if statements from `jit_guard_known_klass` allowing us to immediately return if a check was not needed.
* I'm also happy with gen_checktype, as it's able to dynamically check against the type it wants.
* The change to `jit_rb_obj_not` is the only place I used `known_truthy`, but in the future I could see it being used by `branchif`/`branchunless`

I didn't make changes to `jit_rb_str_concat` (though I think it would be useful there) to avoid conflicts with #6205 cc @noahgibbs 

I'm interested in making some helpers for Type that go in the other direction: from a class or value type to a `Type`, much like `Type::from` does today, but I'd like feedback on this first. That next step could allow introducing new types and have them used (ex. by `jit_guard_known_klass`) without any changes to codegen.

### CODE REVIEW

File: yjit/src/core.rs [hunk 1]
```diff
@@ -126,6 +126,60 @@ impl Type {
...
+            Type::Array => Some(RUBY_T_ARRAY),
+            Type::Hash => Some(RUBY_T_HASH),
+            Type::ImmSymbol | Type::HeapSymbol => Some(RUBY_T_SYMBOL),
+            Type::TString | Type::CString => Some(RUBY_T_STRING),
+            Type::Unknown | Type::UnknownImm | Type::UnknownHeap => None
+        }
+    }
+
+    /// Returns an Option with the class if it is known, otherwise None
+    pub fn known_class_of(&self) -> Option<VALUE> {
```
Comment (maximecb) [inline yjit/src/core.rs:146]: I think I would just rename this to `known_class`, which would be consistent with `known_exact_value` the "class_of" makes me think "is it a subclass of some other thing?".

File: yjit/src/core.rs [hunk 2]
```diff
@@ -126,6 +126,60 @@ impl Type {
...
+        match self {
+            Type::Nil => Some(Qnil),
+            Type::True => Some(Qtrue),
+            Type::False => Some(Qfalse),
+            _ => None,
+        }
+    }
+
+    /// Returns an Option with the exact value if it is known, otherwise None
+    pub fn known_truthy(&self) -> Option<bool> {
```
Comment (maximecb) [inline yjit/src/core.rs:173]: I'd rename this one to `known_bool_value` for consistency.

**Review by maximecb — COMMENTED** [review]
This seems like a nice improvement.

I added some really minor comments wrt naming but I feel like this is clearly worth merging.

File: yjit/src/core.rs (hunk 1 above)
Comment (jhawthorn) [reply yjit/src/core.rs:146]: 👍 `known_class` sounds better to me too. I also wondered if we should use `known_klass` as that's the spelling we used in `codegen.rs`, but I do prefer `known_class` 😓.

File: yjit/src/core.rs (hunk 2 above)
Comment (jhawthorn) [reply yjit/src/core.rs:173]: I'd still prefer "truthy", it's not really a "bool value" (though it does return a bool value 😅) but an emulation of how Ruby's `if` statement behaves. Generally we'll refer to this truthy/falsy

**Pull request merged**